### ISSUE TITLE
Testing suggestion

### ISSUE THREAD

**Issue by lurch** [resin-io-modules/mountutils#17]
Maybe some of the issues we've run into with [drivelist](https://github.com/resin-io-modules/drivelist), are also things that should be actively tested with `mountutils`?

**Comment by jviotti** [comment]
I agree. Lets compile a list of things to test here, and we can the move it to a markdown documentation in the repo, as a manual testing guide:

- Drives whose mount points contain spaces
- Drives whose mount points contain quotes (single and double)
- Drives whose mount points contain symbols such as `#`, and `&`
- Unmount an empty internal reader device on Windows

**Comment by lurch** [comment]
@jviotti Do you remember what the issue you had was, where you were getting different behaviour when the SD-reader was plugged in via a USB3 hub, than when it was plugged in directly?

Any manual testing guide should also include the things you've already been testing this module against, e.g. unmounting non-existent drives / devices, unmounting already-unmounted devices, unmounting drives with multiple-mounted partitions, etc. etc.
There's (probably) going to be so many edge cases, that IMHO the more different things we can add to a list of manual tests, the better. When you're not running through a checklist, it's easy to remember to test one thing on one OS, but then forget to test that same thing on a different OS.

I wonder if we could do any (basic) automated Linux testing by maybe scripting Qemu?

**Comment by jviotti** [comment]
> @jviotti Do you remember what the issue you had was, where you were getting different behaviour when the SD-reader was plugged in via a USB3 hub, than when it was plugged in directly?

I'm afraid I don't :( Do you remember if this was mentioned in a GH issue, or in the Gitter channel?

> Any manual testing guide should also include the things you've already been testing this module against, e.g. unmounting non-existent drives / devices, unmounting already-unmounted devices, unmounting drives with multiple-mounted partitions, etc. etc.
> There's (probably) going to be so many edge cases, that IMHO the more different things we can add to a list of manual tests, the better. When you're not running through a checklist, it's easy to remember to test one thing on one OS, but then forget to test that same thing on a different OS.

I agree, let me update the list above with what I've been testing so far.

> I wonder if we could do any (basic) automated Linux testing by maybe scripting Qemu?

Sounds interesting. What would we automatically mount in the VM though? Can we somehow fake a drive?

**Comment by lurch** [comment]
> I'm afraid I don't :( Do you remember if this was mentioned in a GH issue, or in the Gitter channel?

Don't remember. I tried searching both of those and can't find anything relevant :(

> I wonder if we could do any (basic) automated Linux testing by maybe scripting Qemu?

> > Sounds interesting. What would we automatically mount in the VM though? Can we somehow fake a drive?

Yup, in theory Qemu should allow you to fake anything ;-)  A quick search finds https://wiki.ubuntu.com/QemuDiskHotplug

**Comment by lurch** [comment]
Should we also test unmounting drives which have open file-handles? I guess it should report "drive is busy" or something on all OSes?

**Comment by jviotti** [comment]
> Yup, in theory Qemu should allow you to fake anything ;-) A quick search finds https://wiki.ubuntu.com/QemuDiskHotplug

Sounds exciting. I'm happy to experiment with it.

> Should we also test unmounting drives which have open file-handles? I guess it should report "drive is busy" or something on all OSes?

Good catch. In fact, all OSes will try to close the open handles before unmounting, and only fail if they can't do so.

**Comment by lurch** [comment]
I guess an easy way to test the un-closable handles behaviour would be to try running `mountutils` from the drive we're actually trying to unmount! ;-)

**Comment by jviotti** [comment]
How evil! :) I'll add that to the list as well.
